modules/apps/27-interchain-accounts/types/codec_test.go

```diff
@@ -156,3 +307,172 @@ func (suite *TypesTestSuite) TestDeserializeAndSerializeCosmosTxWithAmino() {
+						Amount:      sdk.NewCoins(sdk.NewCoin("bananas", sdkmath.NewInt(100))),
+					},
+				}
+			},
+			true,
+		},
+		{
+			"success: multiple msgs, different types",
+			func() {
+				cwBytes = []byte(`{
+					"messages": [
+						{
+							"@type": "/cosmos.bank.v1beta1.MsgSend",
+							"from_address": "cosmos17dtl0mjt3t77kpuhg2edqzjpszulwhgzuj9ljs",
+							"to_address": "cosmos17dtl0mjt3t77kpuhg2edqzjpszulwhgzuj9ljs",
+							"amount": [{ "denom": "bananas", "amount": "100" }]
+						},
+						{
+							"@type": "/cosmos.staking.v1beta1.MsgDelegate",
+							"delegator_address": "cosmos17dtl0mjt3t77kpuhg2edqzjpszulwhgzuj9ljs",
+							"validator_address": "cosmos17dtl0mjt3t77kpuhg2edqzjpszulwhgzuj9ljs",
+							"amount": { "denom": "stake", "amount": "5000" }
+						}
+					]
+				}`)
+				protoMessages = []proto.Message{
+					&banktypes.MsgSend{
+						FromAddress: TestOwnerAddress,
+						ToAddress:   TestOwnerAddress,
+						Amount:      sdk.NewCoins(sdk.NewCoin("bananas", sdkmath.NewInt(100))),
+					},
+					&stakingtypes.MsgDelegate{
+						DelegatorAddress: TestOwnerAddress,
+						ValidatorAddress: TestOwnerAddress,
+						Amount:           sdk.NewCoin(sdk.DefaultBondDenom, sdkmath.NewInt(5000)),
+					},
+				}
+			},
+			true,
+		},
+		{
+			"failure: unregistered msg type",
+			func() {
+				cwBytes = []byte(`{"messages":[{}]}`)
+				protoMessages = []proto.Message{
+					&mockSdkMsg{},
+				}
+			},
+			false,
+		},
+		{
+			"failure: multiple unregistered msg types",
+			func() {
+				cwBytes = []byte(`{"messages":[{},{},{}]}`)
+				protoMessages = []proto.Message{
+					&mockSdkMsg{},
+					&mockSdkMsg{},
+					&mockSdkMsg{},
+				}
+			},
+			false,
+		},
+		{
+			"failure: empty bytes",
+			func() {
+				cwBytes = []byte{}
+			},
+			false,
+		},
+	}
+
+	for _, tc := range testCases {
+		tc := tc
+
+		suite.Run(tc.name, func() {
+			tc.malleate()
+			msgs, errDeserialize := types.DeserializeCosmosTx(simapp.MakeTestEncodingConfig().Codec, cwBytes, types.EncodingJSON)
+			if tc.expPass {
+				suite.Require().NoError(errDeserialize, tc.name)
+				for i, msg := range msgs {
+					suite.Require().Equal(protoMessages[i], msg)
+				}
+			} else {
+				suite.Require().Error(errDeserialize, tc.name)
+			}
+		})
+	}
+}
+
+func (suite *TypesTestSuite) TestUnsupportedEncodingType() {
+	// Test serialize
+	msgs := []proto.Message{
+		&banktypes.MsgSend{
+			FromAddress: TestOwnerAddress,
+			ToAddress:   TestOwnerAddress,
+			Amount:      sdk.NewCoins(sdk.NewCoin("bananas", sdkmath.NewInt(100))),
+		},
+	}
+
+	bz, err := types.SerializeCosmosTx(simapp.MakeTestEncodingConfig().Codec, msgs, "unsupported")
+	suite.Require().Error(err)
+	suite.Require().Nil(bz)
+
+	data, err := types.SerializeCosmosTx(simapp.MakeTestEncodingConfig().Codec, msgs, types.EncodingProtobuf)
+	suite.Require().NoError(err)
+
+	_, err = types.DeserializeCosmosTx(simapp.MakeTestEncodingConfig().Codec, data, "unsupported")
+	suite.Require().Error(err)
```

Comment:
probably makes sense to show it'd pass otherwise:
```suggestion
	data, err := types.SerializeCosmosTx(simapp.MakeTestEncodingConfig().Codec, msgs, types.EncodingProtobuf)
	suite.Require().NoError(err)

	_, err = types.DeserializeCosmosTx(simapp.MakeTestEncodingConfig().Codec, data, "unsupported")
	suite.Require().Error(err)
	
	_, err = types.DeserializeCosmosTx(simapp.MakeTestEncodingConfig().Codec, data, types.EncodingProtobuf)
	suite.Require().NoError(err)
```

Reply:
done